aten/src/ATen/NamedTensor.cpp

```diff
@@ -12,6 +12,37 @@ bool NamedTensorMeta::has_names() const {
       });
 }
 
+static DimnameList::const_iterator find_untagged_name(
+    DimnameList::const_iterator begin,
+    DimnameList::const_iterator end,
+    Symbol target) {
+  return std::find_if(begin, end,
+      [&target](const Dimname& candidate) { return candidate.untagged_name() == target; });
+}
+
+static void check_unique_names(DimnameList names) {
+  // Strategy: Compare each element with the ones that come after it.
+  // Although this is O(N^2), in practice N is small (no more than 25).
+  for (auto it = names.begin(); it != names.end(); ++it) {
+    if (it->type() == NameType::WILDCARD) {
+      continue;
+    }
+    auto dup = find_untagged_name(it + 1, names.end(), it->untagged_name());
+    while (dup != names.end()) {
+      TORCH_CHECK(it->name() != dup->name(),
+          "Cannot construct a tensor with duplicate names. Got names: ",
+          names, ".");
+
+      // "C.in" and "C" are not allowed, but "C.in" and "C.out" are.
+      TORCH_CHECK(it->type() == NameType::TAGGED && dup->type() == NameType::TAGGED,
+          "Cannot construct a tensor with duplicate names unless they are tagged ",
+          "and have different tags. Got names: ", names, ", offending names: (",
+          *it, " and ", *dup, ").");
```

Comment:
Will this catch `names=(C.in, C.in)`? it seems like it only checks that both are tagged, not that the tags are unique.

Reply:
The condition above will catch `names=(C.in, C.in)` and complain that the user is constructing a tensor with duplicate names

Reply:
ah I missed that that's the full name.

Would it be better if Dimname attributes were `name` and `tag`, instead of `name` and `untagged_name`? Is  the full name is used more frequently?
Will there be cases where we need to extract the tag in isolation, meaning needing to parse the string from the name symbol?

Reply:
I don't see a case where we need to use the tag in isolation. The full name is going to be used pretty frequently (in `unify(Dimname, Dimname`).

I agree that "name", "tag" and "full name" get confusing to reason about. I don't know what a better naming scheme would be, though.

Reply:
Do you think it would be clearer if the Dimname attributes were `name` and `fullname` instead of `untagged_name` and `name`?

Reply:
I'm going to rename Dimname::name to Dimname::full_name for readability but keep `Dimname::untagged_name`